File: html/en/7gui/flight_booker.c
#include <stdlib.h>
#include <iup.h>

int bt_book_cb(Ihandle *self)
{
  Ihandle* list = IupGetDialogChild(self, "LIST");
  Ihandle* startDate = IupGetDialogChild(self, "STARTDATE");
  Ihandle* endDate = IupGetDialogChild(self, "ENDDATE");

  int flightType = IupGetInt(list, "VALUE");

  if (flightType == 1)
    IupMessagef("Attention!", "You have booked a one-way flight on %s.", IupGetAttribute(startDate, "VALUE"));
  else
    IupMessagef("Attention!", "You have booked a return flight on %s"
                              " and %s.", IupGetAttribute(startDate, "VALUE"), IupGetAttribute(endDate, "VALUE"));

  return IUP_DEFAULT;
}

int validateDate(int day, int month, int year)
{
  int leapYear = 0;

  if (day < 1)
    return 1;

  if (year % 400 == 0 || year % 4 == 0 && year % 100 != 0)
    leapYear = 1;

  if (month == 2 && leapYear && day > 29)
    return 1;

  if (month == 2 && !leapYear && day > 28)
    return 1;

  if ((month == 4 || month == 6 || month == 9 || month == 11) && day > 30)
    return 1;

  if ((month == 1 || month == 3 || month == 5 || month == 7 ||
    month == 8 || month == 10 || month == 12) && day > 31)
    return 1;

  return 0;
}

int checkDate(Ihandle *self)
{
  int year, month, day;
  char *value = IupGetAttribute(self, "VALUE");
  char *part, *ret;
  int count = 1, notValid = 0;

  part = strtok(value, ".");
  if (part == NULL)
    return 1;
  while (part != NULL)
  {
    int len = strlen(part);
    switch (count)
    {
      case 1:
        if (strlen(part) > 2)
          notValid = 1;
        else
        {
          day = strtol(part, &ret, 10);
          if (*ret != '\0')
            notValid = 1;
        }
        break;
      case 2:
        if (strlen(part) > 2)
          notValid = 1;
        else
        {
          month = strtol(part, &ret, 10);
          if (*ret != '\0')
            notValid = 1;
        }
        break;
      case 3:
        if (strlen(part) != 4)
          notValid = 1;
        else
        {
          year = strtol(part, &ret, 10);
          if (*ret != '\0')
            notValid = 1;
        }
        break;
      default:
        notValid = 1;
    }
    count++;
    part = strtok(NULL, ".");
  }

  if (count < 4 || notValid)
    return 1;

  notValid = validateDate(day, month, year);

  if (notValid)
    IupSetAttribute(self, "VALID", "NO");
  else
  {
    IupSetAttribute(self, "VALID", "YES");
    IupSetInt(self, "DAY", day);
    IupSetInt(self, "MONTH", month);
    IupSetInt(self, "YEAR", year);
  }

  return notValid;
}

int startBeforeEnd(Ihandle *self)
{
  Ihandle* startDate = IupGetDialogChild(self, "STARTDATE");
  Ihandle* endDate = IupGetDialogChild(self, "ENDDATE");
  int startDay, startMonth, startYear;
  int endDay, endMonth, endYear;

  if (!IupGetInt(endDate, "ACTIVE"))
    return 1;

  startYear = IupGetInt(startDate, "YEAR");
  endYear = IupGetInt(endDate, "YEAR");

  if (startYear < endYear)
    return 1;

  startMonth = IupGetInt(startDate, "MONTH");
  endMonth = IupGetInt(endDate, "MONTH");

  if (startYear == endYear && startMonth < endMonth)
    return 1;

  startDay = IupGetInt(startDate, "DAY");
  endDay = IupGetInt(endDate, "DAY");

  if (startYear == endYear && startMonth == endMonth && startDay <= endDay)
    return 1;

  return 0;
}

int txt_valuechanged_cb(Ihandle *self)
{
  Ihandle *other = (Ihandle *)IupGetAttribute(self, "OTHER");
  Ihandle* button = IupGetDialogChild(self, "BUTTON");
  int notValid = checkDate(self);

  if (!notValid)
    IupSetAttribute(self, "BGCOLOR", "255 255 255");
  else
    IupSetAttribute(self, "BGCOLOR", "255 0 0");

  if (!notValid && (IupGetInt(other, "VALID") || !IupGetInt(other, "ACTIVE")) &&
      startBeforeEnd(self))
    IupSetAttribute(button, "ACTIVE", "YES");
  else
    IupSetAttribute(button, "ACTIVE", "NO");

  return IUP_DEFAULT;
}

int list_action_cb(Ihandle *self, char *text, int item, int state)
{
  Ihandle* startDate = IupGetDialogChild(self, "STARTDATE");
  Ihandle* endDate = IupGetDialogChild(self, "ENDDATE");
  Ihandle* button = IupGetDialogChild(self, "BUTTON");

  if (state == 0)
    return IUP_DEFAULT;

  if (item == 1)
    IupSetAttribute(endDate, "ACTIVE", "NO");
  else
    IupSetAttribute(endDate, "ACTIVE", "YES");

  if (item==2 && !startBeforeEnd(self))
      IupSetAttribute(button, "ACTIVE", "NO");

  return IUP_DEFAULT;
}

int main(int argc, char **argv)
{
  Ihandle *dlg, *vbox, *list, *startDate, *endDate, *button;

  IupOpen(&argc, &argv);

  list = IupList(NULL);

  IupSetAttribute(list, "NAME", "LIST");
  IupSetAttribute(list, "EXPAND", "HORIZONTAL");
  IupSetAttribute(list, "DROPDOWN", "YES");
  IupSetAttribute(list, "1", "one-way flight");
  IupSetAttribute(list, "2", "return flight");
  IupSetAttribute(list, "VALUE", "1");

  startDate = IupText(NULL);
  endDate = IupText(NULL);

  IupSetAttribute(startDate, "EXPAND", "HORIZONTAL");
  IupSetAttribute(startDate, "NAME", "STARTDATE");
  IupSetAttribute(startDate, "OTHER", (char *)endDate);
  IupSetAttribute(startDate, "VALUE", "22.09.1957");
  IupSetAttribute(startDate, "DAY", "22");
  IupSetAttribute(startDate, "MONTH", "09");
  IupSetAttribute(startDate, "YEAR", "1957");
  IupSetAttribute(startDate, "VALID", "YES");
  IupSetAttribute(endDate, "EXPAND", "HORIZONTAL");
  IupSetAttribute(endDate, "NAME", "ENDDATE");
  IupSetAttribute(endDate, "OTHER", (char *)startDate);
  IupSetAttribute(endDate, "VALUE", "22.09.1957");
  IupSetAttribute(endDate, "VALID", "YES");
  IupSetAttribute(endDate, "DAY", "22");
  IupSetAttribute(endDate, "MONTH", "09");
  IupSetAttribute(endDate, "YEAR", "1957");
  IupSetAttribute(endDate, "ACTIVE", "NO");

  button = IupButton("Book", NULL);

  IupSetAttribute(button, "EXPAND", "HORIZONTAL");
  IupSetAttribute(button, "NAME", "BUTTON");

  vbox = IupVbox(
    list,
    startDate,
    endDate,
    button,
    NULL);
  IupSetAttribute(vbox, "MARGIN", "10x10");
  IupSetAttribute(vbox, "GAP", "10");
  dlg = IupDialog(vbox);
  IupSetAttribute(dlg, "TITLE", "Book Flight");
  IupSetAttribute(dlg, "SIZE", "150");

  /* Registers callbacks */
  IupSetCallback(startDate, "VALUECHANGED_CB", (Icallback)txt_valuechanged_cb);
  IupSetCallback(endDate, "VALUECHANGED_CB", (Icallback)txt_valuechanged_cb);
  IupSetCallback(list, "ACTION", (Icallback)list_action_cb);
  IupSetCallback(button, "ACTION", (Icallback)bt_book_cb);

  IupShowXY(dlg, IUP_CENTER, IUP_CENTER);

  IupMainLoop();

  IupClose();
  return EXIT_SUCCESS;
}

File: html/en/7gui/flight_booker.lua
require("iuplua")

function checkDateFormat(str_date)
	str_date = string.gsub(str_date, "^%s*(.-)%s*$", "%1")
	local check = string.find(str_date, "^%d?%d.%d?%d.%d%d%d%d$")
	if not check then
		return nil
	end
	local day, month, year = string.match(str_date, "(%d?%d).(%d?%d).(%d%d%d%d)")
	if day==nil or month==nil or year==nil then
		return nil
	end
	return 1, tonumber(day), tonumber(month), tonumber(year)
end

function validateDate(day, month, year)
	local leapYear = 0

	if day and day < 1 or day>31 or month and month < 1 or month > 12 then
		return nil
	end

	if year % 400 == 0 or year % 4 == 0 and year % 100 ~= 0 then
		leapYear = 1
	end

	if month == 2 and leapYear and day > 29 then
		return nil
	end

	if month == 2 and not leapYear and day > 28 then
		return nil
	end

	if (month == 4 or month == 6 or month == 9 or month == 11) and day > 30 then
		return nil
	end

  return 1
end

function startBeforeEnd(startDay, startMonth, startYear, endDay, endMonth, endYear)

  if startYear < endYear then
    return 1
  end

  if  startYear == endYear and startMonth < endMonth then
    return 1
  end

  if  startYear == endYear and startMonth == endMonth and startDay <= endDay then
    return 1
  end

  return nil
end

--********************************** Main *****************************************

lst_flight = iup.list{"one-way flight", "return flight"; dropdown = "YES", value = 1, expand = "HORIZONTAL"}
txt_startDate = iup.text{expand = "HORIZONTAL", value = "22.09.1957"}
txt_returnDate = iup.text{expand = "HORIZONTAL", active="NO", value = "22.09.1957"}
btn_book = iup.button{title = "Book", expand = "HORIZONTAL"}

dlg = iup.dialog{iup.vbox{lst_flight, txt_startDate, txt_returnDate, btn_book; gap="10"}, title = "Book Flight", size = "150", margin = "10x10"}

function btn_book:action()
  local flightType = lst_flight.value
  if lst_flight.value == "1" then
    iup.Message("Atention!", "You have booked a one-way flight on "..txt_startDate.value..".")
  else
    iup.Message("Atention!", "You have booked a return flight on "..txt_startDate.value..
                            " and "..txt_returnDate.value..".")
 end
end

function lst_flight:valuechanged_cb()
	if self.value == "1" then
		txt_returnDate.active = "NO"
	else
		txt_returnDate.active = "YES"
	end
end

function checkDate(self)
	local check, day, month, year = checkDateFormat(self.value)
	if  not check then
		self.bgcolor = "255 0 0"
		self.valid = nil
		return
	end
	
	local valid = validateDate(day, month, year)
	if  not valid then
		self.bgcolor = "255 0 0"
		self.valid = nil
		return
	end

	self.bgcolor = "255 255 255"
	self.valid = 1
	self.day = day
	self.month = month
	self.year = year
end

function txt_startDate:valuechanged_cb()
	checkDate(self)
	if self.valid and txt_returnDate.active == "NO" or
	   (txt_returnDate.valid and startBeforeEnd(self.day, self.month, self.year,
		                                        txt_returnDate.day, txt_returnDate.month, txt_returnDate.year)) then
		btn_book.active = "YES"
	else
		btn_book.active = "NO"
	end
end

function txt_returnDate:valuechanged_cb()
	checkDate(self)
	if self.valid and txt_startDate.valid and
	   startBeforeEnd(txt_startDate.day, txt_startDate.month, txt_startDate.year,
		              self.day, self.month, self.year) then
		btn_book.active = "YES"
	else
		btn_book.active = "NO"
	end
end

dlg:showxy( iup.CENTER, iup.CENTER )

if (iup.MainLoopLevel()==0) then
  iup.MainLoop()
end
